src/NuGet.Core/NuGet.Packaging/PackageExtraction/ZipArchiveExtensions.cs

```diff
@@ -67,25 +68,65 @@ public static string SaveAsFile(this ZipArchiveEntry entry, string fileFullPath,
 
         public static void UpdateFileTimeFromEntry(this ZipArchiveEntry entry, string fileFullPath, ILogger logger)
         {
-            var attr = File.GetAttributes(fileFullPath);
+            Testable.Default.UpdateFileTimeFromEntry(entry, fileFullPath, logger);
+        }
+
+        internal class Testable
+        {
+            public static Testable Default { get; } = new Testable(EnvironmentVariableWrapper.Instance);
 
-            if (!attr.HasFlag(FileAttributes.Directory) &&
-                entry.LastWriteTime.DateTime != DateTime.MinValue && // Ignore invalid times
-                entry.LastWriteTime.UtcDateTime <= DateTime.UtcNow) // Ignore future times
+            internal Testable(IEnvironmentVariableReader environmentVariableReader)
             {
-                try
+                _updateFileTimeFromEntryMaxRetries = 9;
+                string value = environmentVariableReader.GetEnvironmentVariable("NUGET_UpdateFileTime_MaxRetries");
+                if (int.TryParse(value, out int maxRetries) && maxRetries > 0)
                 {
-                    File.SetLastWriteTimeUtc(fileFullPath, entry.LastWriteTime.Add(entry.LastWriteTime.Offset).UtcDateTime);
+                    _updateFileTimeFromEntryMaxRetries = maxRetries;
                 }
-                catch (ArgumentOutOfRangeException ex)
+            }
+
+            private readonly int _updateFileTimeFromEntryMaxRetries;
+
+            public void UpdateFileTimeFromEntry(ZipArchiveEntry entry, string fileFullPath, ILogger logger)
+            {
+                if (entry == null) throw new ArgumentNullException(nameof(entry));
+                if (fileFullPath == null) throw new ArgumentNullException(nameof(fileFullPath));
+                if (logger == null) throw new ArgumentNullException(nameof(logger));
+
+                var attr = File.GetAttributes(fileFullPath);
```

Comment:
Maybe I might be wrong, but this one making actual I/O read. What would be perf cost for additional I/O call per every zip entry? Do we really need it?

Reply:
well, `attr` is used on the next line to make sure it's not a directory, which suggests that this method is sometimes called on directories. `File.SetLastWriteTime` will throw if the path is a directory. So my assumption is yes, it's needed.

Having said that, it might be less costly if we check if the path ends in a slash, rather than doing disk IO, but I'm not sure that it would always work the same way. In fact, I can think of a way to create an intentially malicious nupkg that would cause such an assumption to fail, so the trailing slash is probably not a safe way to ensure the path is a file. Having said that I expect that the file copy before this method is called would fail in such a case.  All this to say that maybe we can reduce disk IO but it's risky and probably needs a lot more testing than what the scope of this PR/issue warrants.

On the other hand, the only reason this is coming up in the diff is because it's indentation changed due to the internal class. If you view the diff ignoring whitespace (or check the red part of the diff) you'll see this is existing code. Not that is justification to avoid making improvements, but it's difficult for me to answer if we really need it given it's code that existed since before I joined the team.